File: lib/python/openlinktoken/src/main/openlinktoken/exchange_kem.py
# SPDX-License-Identifier: MIT
"""Version-2 exchange facade backed by standard JWE JSON Serialization."""

from __future__ import annotations

import base64
import json
from typing import Any, Mapping

from openlinktoken.crypto_suite import CryptoSuite
from openlinktoken.exchange_key_bundle import ExchangeKeyBundle
from openlinktoken.jwe_mlkem import (
    EXCHANGE_V2_CONTENT_TYPE,
    EXCHANGE_V2_ENCRYPTION,
    EXCHANGE_V2_TYPE,
    EXCHANGE_V2_VERSION,
    build_v2_jwe,
    decrypt_v2_jwe,
)


def build_exchange_envelope_v2(
    exchange_name: str,
    hashing_secret: bytes,
    sender_bundle: ExchangeKeyBundle,
    recipient_bundle: ExchangeKeyBundle,
    created_at: str,
    exchange_id: str,
    rotation_iv: bytes = b"",
    rotation_count: int = 0,
    bin_width: float = 0.05,
    dimension_bias: list[float] | None = None,
) -> dict[str, Any]:
    """Build a version-2 exchange config as a standard general JWE JSON object."""
    if sender_bundle.suite != recipient_bundle.suite:
        raise ValueError("Sender and recipient key bundles must use the same crypto suite.")
    suite = sender_bundle.suite
    if suite.exchange_config_version != EXCHANGE_V2_VERSION:
        raise ValueError(f"Suite '{suite.suite_id}' does not use exchange configuration version 2.")
    if not exchange_name or not exchange_id:
        raise ValueError("Exchange name and exchange ID must be non-empty.")
    if not isinstance(hashing_secret, bytes):
        raise TypeError("Hashing secret must be bytes.")
    if not isinstance(rotation_iv, bytes):
        raise TypeError("Rotation IV must be bytes.")
    if rotation_count < 0:
        raise ValueError("Rotation count must be non-negative.")
    if bin_width <= 0:
        raise ValueError("Bin width must be positive.")

    protected_header = {
        "typ": EXCHANGE_V2_TYPE,
        "cty": EXCHANGE_V2_CONTENT_TYPE,
        "enc": EXCHANGE_V2_ENCRYPTION,
        "version": EXCHANGE_V2_VERSION,
        "cryptoSuite": suite.suite_id,
        "exchangeId": exchange_id,
    }
    payload = {
        "exchangeName": exchange_name,
        "cryptoSuite": suite.suite_id,
        "hashingSecret": _encode(hashing_secret),
        "hashingSecretEncoding": "base64url",
        "senderKeyId": sender_bundle.kid,
        "recipientKeyId": recipient_bundle.kid,
        "senderKeyBundle": sender_bundle.to_mapping(),
        "recipientKeyBundle": recipient_bundle.to_mapping(),
        "createdAt": created_at,
        "exchangeId": exchange_id,
        "rotationIv": _encode(rotation_iv),
        "rotationIvEncoding": "base64url",
        "rotationCount": rotation_count,
        "binWidth": bin_width,
        "dimensionBias": dimension_bias if dimension_bias is not None else [],
    }
    return build_v2_jwe(
        _canonical_json(payload),
        protected_header,
        [sender_bundle, recipient_bundle],
    )


def decrypt_exchange_envelope_v2(
    exchange_config: Mapping[str, Any],
    private_bundle_value: bytes | str | Mapping[str, Any] | ExchangeKeyBundle,
) -> tuple[bytes, bytes]:
    """Decrypt a standard version-2 exchange config and return plaintext plus transport key."""
    private_bundle = _parse_private_bundle(private_bundle_value)
    plaintext, transport_key = decrypt_v2_jwe(exchange_config, private_bundle)
    protected_header = _decode_protected_header(exchange_config.get("protected"))
    suite = CryptoSuite.from_id(protected_header["cryptoSuite"])
    payload = _parse_payload(plaintext)
    _validate_payload(payload, suite, protected_header)
    return plaintext, transport_key


def decrypt_exchange_envelope(
    exchange_config: Mapping[str, Any],
    private_bundle_value: bytes | str | Mapping[str, Any] | ExchangeKeyBundle,
) -> bytes:
    """Decrypt a version-2 exchange config and return only its plaintext."""
    plaintext, _ = decrypt_exchange_envelope_v2(exchange_config, private_bundle_value)
    return plaintext


def _parse_private_bundle(
    value: bytes | str | Mapping[str, Any] | ExchangeKeyBundle,
) -> ExchangeKeyBundle:
    """Parse and validate private version-2 key material."""
    if isinstance(value, ExchangeKeyBundle):
        bundle = value
    elif isinstance(value, Mapping):
        bundle = ExchangeKeyBundle.from_mapping(value, require_private=True)
    else:
        bundle = ExchangeKeyBundle.from_json(value, require_private=True)
    return bundle


def _parse_payload(value: bytes) -> dict[str, Any]:
    """Parse the decrypted exchange payload as a JSON object."""
    try:
        payload = json.loads(value)
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        raise ValueError(f"Version-2 exchange payload is not valid JSON: {error}") from error
    if not isinstance(payload, dict):
        raise ValueError("Version-2 exchange payload must be a JSON object.")
    return payload


def _validate_payload(
    payload: Mapping[str, Any],
    suite: CryptoSuite,
    protected_header: Mapping[str, Any],
) -> None:
    """Validate payload identity fields against the authenticated protected header."""
    if payload.get("cryptoSuite") != suite.suite_id:
        raise ValueError("Version-2 exchange payload suite does not match the protected header.")
    if payload.get("exchangeId") != protected_header.get("exchangeId"):
        raise ValueError("Version-2 exchange payload exchangeId does not match the protected header.")

    for field_name in ("senderKeyId", "recipientKeyId"):
        if not isinstance(payload.get(field_name), str) or not payload[field_name]:
            raise ValueError(f"Version-2 exchange payload is missing {field_name}.")

    for field_name, expected_kid in (
        ("senderKeyBundle", payload["senderKeyId"]),
        ("recipientKeyBundle", payload["recipientKeyId"]),
    ):
        bundle_value = payload.get(field_name)
        if not isinstance(bundle_value, Mapping):
            raise ValueError(f"Version-2 exchange payload is missing {field_name}.")
        bundle = ExchangeKeyBundle.from_mapping(bundle_value)
        if bundle.suite != suite or bundle.kid != expected_kid:
            raise ValueError(f"Version-2 exchange payload {field_name} does not match its key ID or suite.")


def _decode_protected_header(value: Any) -> dict[str, Any]:
    """Decode the protected header needed for payload consistency validation."""
    protected_bytes = _decode(value, "protected")
    try:
        header = json.loads(protected_bytes)
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        raise ValueError(f"Protected version-2 exchange header is not valid JSON: {error}") from error
    if not isinstance(header, dict):
        raise ValueError("Protected version-2 exchange header must be a JSON object.")
    return header


def _canonical_json(value: Mapping[str, Any]) -> bytes:
    """Serialize a mapping deterministically for the JWE plaintext."""
    return json.dumps(value, sort_keys=True, separators=(",", ":")).encode("utf-8")


def _encode(value: bytes) -> str:
    """Encode bytes as unpadded base64url text."""
    return base64.urlsafe_b64encode(value).decode("ascii").rstrip("=")


def _decode(value: Any, field_name: str) -> bytes:
    """Decode a required base64url field."""
    if not isinstance(value, str) or not value:
        raise ValueError(f"{field_name} must be non-empty base64url data.")
    try:
        padding = "=" * (-len(value) % 4)
        return base64.b64decode(value + padding, altchars=b"-_", validate=True)
    except (ValueError, UnicodeEncodeError) as error:
        raise ValueError(f"{field_name} is not valid base64url data: {error}") from error

File: lib/java/openlinktoken/src/main/java/org/openlinktoken/ExchangeKem.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken;

import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.openlinktoken.crypto.CryptoSuite;

/**
 * Version-two ML-KEM exchange facade backed by standard general JWE JSON serialization.
 */
public final class ExchangeKem {

    /** Version-two exchange envelope version. */
    public static final int VERSION = JweMlkem.EXCHANGE_V2_VERSION;

    /** Version-two protected JWE type. */
    public static final String TYPE = JweMlkem.EXCHANGE_V2_TYPE;

    /** Version-two protected JWE content type. */
    public static final String CONTENT_TYPE = JweMlkem.EXCHANGE_V2_CONTENT_TYPE;

    /** Version-two protected JWE content-encryption algorithm. */
    public static final String ENCRYPTION = JweMlkem.EXCHANGE_V2_ENCRYPTION;

    /** Pure ML-KEM-768 recipient algorithm. */
    public static final String PURE_KEM_ALGORITHM = JweMlkem.PURE_KEM_ALGORITHM;

    /** Hybrid P-256 and ML-KEM-768 recipient algorithm. */
    public static final String HYBRID_KEM_ALGORITHM = JweMlkem.HYBRID_KEM_ALGORITHM;

    /** HKDF domain-separation label for the v2 token transport key. */
    public static final String TOKEN_TRANSPORT_KEY_INFO = JweMlkem.TOKEN_TRANSPORT_KEY_INFO;

    /** Alias matching the versioned Python exchange helper constants. */
    public static final int EXCHANGE_V2_VERSION = VERSION;

    /** Alias matching the versioned Python exchange helper constants. */
    public static final String EXCHANGE_V2_TYPE = TYPE;

    /** Alias matching the versioned Python exchange helper constants. */
    public static final String EXCHANGE_V2_CONTENT_TYPE = CONTENT_TYPE;

    /** Alias matching the versioned Python exchange helper constants. */
    public static final String EXCHANGE_V2_ENCRYPTION = ENCRYPTION;

    private static final String HASHING_SECRET_ENCODING = "base64url";
    private static final String ROTATION_IV_ENCODING = "base64url";

    private ExchangeKem() {
    }

    /**
     * Builds a version-two exchange envelope using the default exchange payload options.
     *
     * @param exchangeName the logical exchange name
     * @param hashingSecret the hashing secret
     * @param senderBundle the sender's private or public key bundle
     * @param recipientBundle the recipient's public key bundle
     * @param createdAt the exchange creation timestamp
     * @param exchangeId the authenticated exchange identifier
     * @return a standard general JWE JSON object
     */
    public static Map<String, Object> buildExchangeEnvelopeV2(
            String exchangeName,
            byte[] hashingSecret,
            ExchangeKeyBundle senderBundle,
            ExchangeKeyBundle recipientBundle,
            String createdAt,
            String exchangeId) {
        return buildExchangeEnvelopeV2(
                exchangeName,
                hashingSecret,
                senderBundle,
                recipientBundle,
                createdAt,
                exchangeId,
                new byte[0],
                0,
                0.05,
                List.of());
    }

    /**
     * Builds a version-two exchange envelope using standard general JWE JSON serialization.
     *
     * @param exchangeName the logical exchange name
     * @param hashingSecret the hashing secret
     * @param senderBundle the sender's private or public key bundle
     * @param recipientBundle the recipient's public key bundle
     * @param createdAt the exchange creation timestamp
     * @param exchangeId the authenticated exchange identifier
     * @param rotationIv the rotation-matrix initialization vector
     * @param rotationCount the number of rotation matrices
     * @param binWidth the quantization bin width
     * @param dimensionBias the rotation bias vector
     * @return a standard general JWE JSON object
     */
    public static Map<String, Object> buildExchangeEnvelopeV2(
            String exchangeName,
            byte[] hashingSecret,
            ExchangeKeyBundle senderBundle,
            ExchangeKeyBundle recipientBundle,
            String createdAt,
            String exchangeId,
            byte[] rotationIv,
            int rotationCount,
            double binWidth,
            List<Double> dimensionBias) {
        if (senderBundle == null || recipientBundle == null) {
            throw new IllegalArgumentException("Sender and recipient key bundles must not be null.");
        }
        if (!senderBundle.getSuite().getSuiteId().equals(recipientBundle.getSuite().getSuiteId())) {
            throw new IllegalArgumentException("Sender and recipient key bundles must use the same crypto suite.");
        }
        CryptoSuite suite = senderBundle.getSuite();
        if (suite.getExchangeConfigVersion() != JweMlkem.EXCHANGE_V2_VERSION) {
            throw new IllegalArgumentException(
                    "Suite '" + suite.getSuiteId() + "' does not use exchange configuration version 2.");
        }
        if (exchangeName == null || exchangeName.isEmpty() || exchangeId == null || exchangeId.isEmpty()) {
            throw new IllegalArgumentException("Exchange name and exchange ID must be non-empty.");
        }
        validateHashingSecret(hashingSecret, suite);
        if (rotationIv == null) {
            throw new IllegalArgumentException("Rotation IV must be bytes.");
        }
        if (rotationCount < 0) {
            throw new IllegalArgumentException("Rotation count must be non-negative.");
        }
        if (!(binWidth > 0.0) || !Double.isFinite(binWidth)) {
            throw new IllegalArgumentException("Bin width must be positive.");
        }
        List<Double> biases = dimensionBias == null ? List.of() : List.copyOf(dimensionBias);

        Map<String, Object> protectedHeader = new LinkedHashMap<>();
        protectedHeader.put("typ", JweMlkem.EXCHANGE_V2_TYPE);
        protectedHeader.put("cty", JweMlkem.EXCHANGE_V2_CONTENT_TYPE);
        protectedHeader.put("enc", JweMlkem.EXCHANGE_V2_ENCRYPTION);
        protectedHeader.put("version", JweMlkem.EXCHANGE_V2_VERSION);
        protectedHeader.put("cryptoSuite", suite.getSuiteId());
        protectedHeader.put("exchangeId", exchangeId);

        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("exchangeName", exchangeName);
        payload.put("cryptoSuite", suite.getSuiteId());
        payload.put("hashingSecret", CryptoEncoding.encodeBase64Url(hashingSecret));
        payload.put("hashingSecretEncoding", HASHING_SECRET_ENCODING);
        payload.put("senderKeyId", senderBundle.getKid());
        payload.put("recipientKeyId", recipientBundle.getKid());
        payload.put("senderKeyBundle", senderBundle.toMapping());
        payload.put("recipientKeyBundle", recipientBundle.toMapping());
        payload.put("createdAt", createdAt);
        payload.put("exchangeId", exchangeId);
        payload.put("rotationIv", CryptoEncoding.encodeBase64Url(rotationIv));
        payload.put("rotationIvEncoding", ROTATION_IV_ENCODING);
        payload.put("rotationCount", rotationCount);
        payload.put("binWidth", binWidth);
        payload.put("dimensionBias", biases);

        return JweMlkem.build(
                JsonSupport.writeObject(payload),
                protectedHeader,
                List.of(senderBundle, recipientBundle));
    }

    /**
     * Decrypts a version-two exchange envelope and returns its payload and transport key.
     *
     * @param exchangeConfig a standard general JWE JSON object
     * @param privateBundle the matching private key bundle
     * @return the authenticated plaintext and the separately derived transport key
     */
    public static DecryptionResult decryptExchangeEnvelopeV2(
            Map<String, ?> exchangeConfig,
            ExchangeKeyBundle privateBundle) {
        JweMlkem.Decryption decrypted = JweMlkem.decrypt(exchangeConfig, privateBundle);
        Map<String, Object> protectedHeader = decrypted.protectedHeader();
        CryptoSuite suite = CryptoSuite.fromId((String) protectedHeader.get("cryptoSuite"));
        Map<String, Object> payload = parsePayload(decrypted.plaintext());
        validatePayload(payload, suite, protectedHeader);
        return new DecryptionResult(decrypted.plaintext(), decrypted.transportKey());
    }

    /**
     * Decrypts a UTF-8 JSON version-two exchange envelope.
     *
     * @param exchangeConfig UTF-8 standard general JWE JSON
     * @param privateBundle the matching private key bundle
     * @return the authenticated plaintext and separately derived transport key
     */
    public static DecryptionResult decryptExchangeEnvelopeV2(
            byte[] exchangeConfig,
            ExchangeKeyBundle privateBundle) {
        try {
            return decryptExchangeEnvelopeV2(JsonSupport.readObject(exchangeConfig), privateBundle);
        } catch (IllegalArgumentException exception) {
            throw exception;
        } catch (RuntimeException exception) {
            throw new IllegalArgumentException("Version-2 exchange envelope is not valid JSON.", exception);
        }
    }

    /**
     * Decrypts a version-two exchange envelope and returns only its plaintext payload.
     *
     * @param exchangeConfig a standard general JWE JSON object
     * @param privateBundle the matching private key bundle
     * @return the authenticated plaintext
     */
    public static byte[] decryptExchangeEnvelope(
            Map<String, ?> exchangeConfig,
            ExchangeKeyBundle privateBundle) {
        return decryptExchangeEnvelopeV2(exchangeConfig, privateBundle).getPlaintext();
    }

    /**
     * Decrypts a UTF-8 JSON version-two exchange envelope and returns its plaintext.
     *
     * @param exchangeConfig UTF-8 standard general JWE JSON
     * @param privateBundle the matching private key bundle
     * @return the authenticated plaintext
     */
    public static byte[] decryptExchangeEnvelope(
            byte[] exchangeConfig,
            ExchangeKeyBundle privateBundle) {
        return decryptExchangeEnvelopeV2(exchangeConfig, privateBundle).getPlaintext();
    }

    private static void validateHashingSecret(byte[] hashingSecret, CryptoSuite suite) {
        if (hashingSecret == null) {
            throw new IllegalArgumentException("Hashing secret must be bytes.");
        }
        if (CryptoSuite.TOKEN_MAC_KMAC256_256.equals(suite.getTokenMacAlgorithm())
                && hashingSecret.length < 32) {
            throw new IllegalArgumentException(
                    "Crypto suite '" + suite.getSuiteId()
                            + "' requires a hashing secret of at least 32 bytes.");
        }
    }

    private static Map<String, Object> parsePayload(byte[] plaintext) {
        try {
            return JsonSupport.readObject(plaintext);
        } catch (IllegalArgumentException exception) {
            throw new IllegalArgumentException(
                    "Version-2 exchange payload is not valid JSON.", exception);
        }
    }

    private static void validatePayload(
            Map<String, Object> payload,
            CryptoSuite suite,
            Map<String, Object> protectedHeader) {
        if (!suite.getSuiteId().equals(payload.get("cryptoSuite"))) {
            throw new IllegalArgumentException(
                    "Version-2 exchange payload suite does not match the protected header.");
        }
        if (!protectedHeader.get("exchangeId").equals(payload.get("exchangeId"))) {
            throw new IllegalArgumentException(
                    "Version-2 exchange payload exchangeId does not match the protected header.");
        }

        String senderKeyId = requiredPayloadString(payload, "senderKeyId");
        String recipientKeyId = requiredPayloadString(payload, "recipientKeyId");
        validatePayloadBundle(payload, "senderKeyBundle", senderKeyId, suite);
        validatePayloadBundle(payload, "recipientKeyBundle", recipientKeyId, suite);
    }

    private static String requiredPayloadString(Map<String, Object> payload, String fieldName) {
        Object value = payload.get(fieldName);
        if (!(value instanceof String string) || string.isEmpty()) {
            throw new IllegalArgumentException("Version-2 exchange payload is missing " + fieldName + ".");
        }
        return string;
    }

    private static void validatePayloadBundle(
            Map<String, Object> payload,
            String fieldName,
            String expectedKid,
            CryptoSuite suite) {
        Object value = payload.get(fieldName);
        if (!(value instanceof Map<?, ?> map)) {
            throw new IllegalArgumentException("Version-2 exchange payload is missing " + fieldName + ".");
        }
        Map<String, Object> mapping = new LinkedHashMap<>();
        for (Map.Entry<?, ?> entry : map.entrySet()) {
            if (!(entry.getKey() instanceof String key)) {
                throw new IllegalArgumentException(fieldName + " must contain string field names.");
            }
            mapping.put(key, entry.getValue());
        }
        ExchangeKeyBundle bundle = ExchangeKeyBundle.fromMapping(mapping);
        if (!suite.getSuiteId().equals(bundle.getSuite().getSuiteId())
                || !expectedKid.equals(bundle.getKid())) {
            throw new IllegalArgumentException(
                    "Version-2 exchange payload " + fieldName + " does not match its key ID or suite.");
        }
    }

    /**
     * Immutable result returned by version-two exchange decryption.
     */
    public static final class DecryptionResult {
        private final byte[] plaintext;
        private final byte[] transportKey;

        private DecryptionResult(byte[] plaintext, byte[] transportKey) {
            this.plaintext = Arrays.copyOf(plaintext, plaintext.length);
            this.transportKey = Arrays.copyOf(transportKey, transportKey.length);
        }

        /**
         * Returns a copy of the decrypted exchange payload.
         *
         * @return UTF-8 JSON payload bytes
         */
        public byte[] getPlaintext() {
            return Arrays.copyOf(plaintext, plaintext.length);
        }

        /**
         * Returns a copy of the decrypted exchange payload.
         *
         * @return UTF-8 JSON payload bytes
         */
        public byte[] plaintext() {
            return getPlaintext();
        }

        /**
         * Returns a copy of the v2 token transport key.
         *
         * @return the 32-byte transport key
         */
        public byte[] getTransportKey() {
            return Arrays.copyOf(transportKey, transportKey.length);
        }

        /**
         * Returns a copy of the v2 token transport key.
         *
         * @return the 32-byte transport key
         */
        public byte[] transportKey() {
            return getTransportKey();
        }
    }
}
